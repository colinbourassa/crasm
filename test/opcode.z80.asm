;;; Author: Leon Bottou
;;; Public Domain.

; All Z80 opcodes

	cpu z80	
	
      *=4000H
    adr=12345	
     dd=23H
      n=45Q
     nn=23456

	adc a,(hl)
	adc a,(ix+dd)
	adc a,(iy+dd)
	adc a,a
	adc a,b
	adc a,c
	adc a,d
	adc a,e
	adc a,h
	adc a,l
	adc a,n
	adc hl,bc
	adc hl,de
	adc hl,hl
	adc hl,sp

	add a,(hl)
	add a,(ix+dd)
	add a,(iy+dd)
	add a,a
	add a,b
	add a,c
	add a,d
	add a,e
	add a,h
	add a,l
	add a,n
	add hl,bc
	add hl,de
	add hl,hl
	add hl,sp
	add ix,bc
	add ix,de
	add ix,ix
	add ix,sp
	add iy,bc
	add iy,de
	add iy,iy
	add iy,sp

	and (hl)
	and (ix+dd)
	and (iy+dd)
	and a
	and b
	and c
	and d
	and e
	and h
	and l
	and n

	bit 0,(hl)
	bit 0,(ix+dd)
	bit 0,(iy+dd)
	bit 0,a
	bit 0,b
	bit 0,c
	bit 0,d
	bit 0,e
	bit 0,h
	bit 0,l
	bit 1,(hl)
	bit 1,(ix+dd)
	bit 1,(iy+dd)
	bit 1,a
	bit 1,b
	bit 1,c
	bit 1,d
	bit 1,e
	bit 1,h
	bit 1,l
	bit 2,(hl)
	bit 2,(ix+dd)
	bit 2,(iy+dd)
	bit 2,a
	bit 2,b
	bit 2,c
	bit 2,d
	bit 2,e
	bit 2,h
	bit 2,l
	bit 3,(hl)
	bit 3,(ix+dd)
	bit 3,(iy+dd)
	bit 3,a
	bit 3,b
	bit 3,c
	bit 3,d
	bit 3,e
	bit 3,h
	bit 3,l
	bit 4,(hl)
	bit 4,(ix+dd)
	bit 4,(iy+dd)
	bit 4,a
	bit 4,b
	bit 4,c
	bit 4,d
	bit 4,e
	bit 4,h
	bit 4,l
	bit 5,(hl)
	bit 5,(ix+dd)
	bit 5,(iy+dd)
	bit 5,a
	bit 5,b
	bit 5,c
	bit 5,d
	bit 5,e
	bit 5,h
	bit 5,l
	bit 6,(hl)
	bit 6,(ix+dd)
	bit 6,(iy+dd)
	bit 6,a
	bit 6,b
	bit 6,c
	bit 6,d
	bit 6,e
	bit 6,h
	bit 6,l
	bit 7,(hl)
	bit 7,(ix+dd)
	bit 7,(iy+dd)
	bit 7,a
	bit 7,b
	bit 7,c
	bit 7,d
	bit 7,e
	bit 7,h
	bit 7,l

	call adr
	call c,adr
	call m,adr
	call nc,adr
	call nz,adr
	call p,adr
	call pe,adr
	call po,adr
	call z,adr

	ccf
	
	cp  (hl)
	cp  (ix+dd)
	cp  (iy+dd)
	cp  a
	cp  b
	cp  c
	cp  d
	cp  e
	cp  h
	cp  l
	cp  n
	
	cpd
	cpdr
	cpi
	cpir
	
	cpl
	daa
	
	dec (hl)
	dec (ix+dd)
	dec (iy+dd)
	dec a
	dec b
	dec bc
	dec c
	dec d
	dec de
	dec e
	dec h
	dec hl
	dec ix
	dec iy
	dec l
	dec sp
	
	di
here1	djnz here1
	ei
	
	ex (sp),hl
	ex (sp),ix
	ex (sp),iy
	ex af,af'
	ex de,hl
	
	exx
	halt
	im 0
	im 1
	im 2
	
	in a,(c)
	in a,(n)
	in b,(c)
	in c,(c)
	in d,(c)
	in e,(c)
	in h,(c)
	in l,(c)
	
	inc (hl)
	inc (ix+dd)
	inc (iy+dd)
	inc a
	inc b
	inc bc
	inc c
	inc d
	inc de
	inc e
	inc h
	inc hl
	inc ix
	inc iy
	inc l
	inc sp
	
	ind
	indr
	ini
	inir
	
	jp (hl)
	jp (ix)
	jp (iy)
	jp adr
	jp c,adr
	jp m,adr
	jp nc,adr
	jp nz,adr
	jp p,adr
	jp pe,adr
	jp po,adr
	jp z,adr
	
	jr c,here2
here2	jr here2
	jr nc,here2
	jr nz,here2
	jr z,here2
	
	ld (bc),a
	ld (de),a
	
	ld (hl),a
	ld (hl),b
	ld (hl),c
	ld (hl),d
	ld (hl),e
	ld (hl),h
	ld (hl),l
	ld (hl),n

	ld (ix+dd),a
	ld (ix+dd),b
	ld (ix+dd),c
	ld (ix+dd),d
	ld (ix+dd),e
	ld (ix+dd),h
	ld (ix+dd),l
	ld (ix+dd),n
	ld (iy+dd),a
	ld (iy+dd),b
	ld (iy+dd),c
	ld (iy+dd),d
	ld (iy+dd),e
	ld (iy+dd),h
	ld (iy+dd),l
	ld (iy+dd),n
	
	ld (nn),a
	ld (nn),bc
	ld (nn),de
	ld (nn),hl
	ld (nn),ix
	ld (nn),iy
	ld (nn),sp
	
	ld a,(bc)
	ld a,(de)
	ld a,(hl)
	ld a,(ix+dd)
	ld a,(iy+dd)
	ld a,(nn)
	ld a,a
	ld a,b
	ld a,c
	ld a,d
	ld a,e
	ld a,h
	ld a,i
	ld a,l
	ld a,n
	ld a,r
	
	ld b,(hl)
	ld b,(ix+dd)
	ld b,(iy+dd)
	ld b,a
	ld b,b
	ld b,c
	ld b,d
	ld b,e
	ld b,h
	ld b,l
	ld b,n
	
	ld bc,(nn)
	ld bc,nn
	
	ld c,(hl)
	ld c,(ix+dd)
	ld c,(iy+dd)
	ld c,a
	ld c,b
	ld c,c
	ld c,d
	ld c,e
	ld c,h
	ld c,l
	ld c,n
	
	ld d,(hl)
	ld d,(ix+dd)
	ld d,(iy+dd)
	ld d,a
	ld d,b
	ld d,c
	ld d,d
	ld d,e
	ld d,h
	ld d,l
	ld d,n
	
	ld de,(nn)
	ld de,nn
	
	ld e,(hl)
	ld e,(ix+dd)
	ld e,(iy+dd)
	ld e,a
	ld e,b
	ld e,c
	ld e,d
	ld e,e
	ld e,h
	ld e,l
	ld e,n
	
	ld h,(hl)
	ld h,(ix+dd)
	ld h,(iy+dd)
	ld h,a
	ld h,b
	ld h,c
	ld h,d
	ld h,e
	ld h,h
	ld h,l
	ld h,n
	
	ld hl,(nn)
	ld hl,nn
	
	ld i,a
	
	ld ix,(nn)
	ld ix,nn
	ld iy,(nn)
	ld iy,nn
	
	ld l,(hl)
	ld l,(ix+dd)
	ld l,(iy+dd)
	ld l,a
	ld l,b
	ld l,c
	ld l,d
	ld l,e
	ld l,h
	ld l,l
	ld l,n
	
	ld r,a
	
	ld sp,(nn)
	ld sp,hl
	ld sp,ix
	ld sp,iy
	ld sp,nn
	
	ldd
	lddr
	ldi
	ldir
	
	neg
	nop
	
	or (hl)
	or (ix+dd)
	or (iy+dd)
	or a
	or b
	or c
	or d
	or e
	or h
	or l
	or n
	
	otdr
	otir
	
	out (c),a
	out (c),b
	out (c),c
	out (c),d
	out (c),e
	out (c),h
	out (c),l
	out (n),a
	
	outd
	outi
	
	pop af
	pop bc
	pop de
	pop hl
	pop ix
	pop iy
	
	push af
	push bc
	push de
	push hl
	push ix
	push iy
	
	res 0,(hl)
	res 0,(ix+dd)
	res 0,(iy+dd)
	res 0,a
	res 0,b
	res 0,c
	res 0,d
	res 0,e
	res 0,h
	res 0,l
	res 1,(hl)
	res 1,(ix+dd)
	res 1,(iy+dd)
	res 1,a
	res 1,b
	res 1,c
	res 1,d
	res 1,e
	res 1,h
	res 1,l
	res 2,(hl)
	res 2,(ix+dd)
	res 2,(iy+dd)
	res 2,a
	res 2,b
	res 2,c
	res 2,d
	res 2,e
	res 2,h
	res 2,l
	res 3,(hl)
	res 3,(ix+dd)
	res 3,(iy+dd)
	res 3,a
	res 3,b
	res 3,c
	res 3,d
	res 3,e
	res 3,h
	res 3,l
	res 4,(hl)
	res 4,(ix+dd)
	res 4,(iy+dd)
	res 4,a
	res 4,b
	res 4,c
	res 4,d
	res 4,e
	res 4,h
	res 4,l
	res 5,(hl)
	res 5,(ix+dd)
	res 5,(iy+dd)
	res 5,a
	res 5,b
	res 5,c
	res 5,d
	res 5,e
	res 5,h
	res 5,l
	res 6,(hl)
	res 6,(ix+dd)
	res 6,(iy+dd)
	res 6,a
	res 6,b
	res 6,c
	res 6,d
	res 6,e
	res 6,h
	res 6,l
	res 7,(hl)
	res 7,(ix+dd)
	res 7,(iy+dd)
	res 7,a
	res 7,b
	res 7,c
	res 7,d
	res 7,e
	res 7,h
	res 7,l

	ret
	ret c
	ret m
	ret nc
	ret nz
	ret p
	ret pe
	ret po
	ret z
	reti
	retn
	
	rl (hl)
	rl (ix+dd)
	rl (iy+dd)
	rl a
	rl b
	rl c
	rl d
	rl e
	rl h
	rl l
	
	rla
	
	rlc (hl)
	rlc (ix+dd)
	rlc (iy+dd)
	rlc a
	rlc b
	rlc c
	rlc d
	rlc e
	rlc h
	rlc l
	
	rlca
	rld
	
	rr (hl)
	rr (ix+dd)
	rr (iy+dd)
	rr a
	rr b
	rr c
	rr d
	rr e
	rr h
	rr l
	
	rra
	
	rrc (hl)
	rrc (ix+dd)
	rrc (iy+dd)
	rrc a
	rrc b
	rrc c
	rrc d
	rrc e
	rrc h
	rrc l
	
	rrca
	rrd
	
	rst 00H
	rst 08H
	rst 10H
	rst 18H
	rst 20H
	rst 28H
	rst 30H
	rst 38H

	sbc a,(hl)
	sbc a,(ix+dd)
	sbc a,(iy+dd)
	sbc a,a
	sbc a,b
	sbc a,c
	sbc a,d
	sbc a,e
	sbc a,h
	sbc a,l
	sbc a,n
	sbc hl,bc
	sbc hl,de
	sbc hl,hl
	sbc hl,sp
	
	scf
	
	set 0,(hl)
	set 0,(ix+dd)
	set 0,(iy+dd)
	set 0,a
	set 0,b
	set 0,c
	set 0,d
	set 0,e
	set 0,h
	set 0,l
	set 1,(hl)
	set 1,(ix+dd)
	set 1,(iy+dd)
	set 1,a
	set 1,b
	set 1,c
	set 1,d
	set 1,e
	set 1,h
	set 1,l
	set 2,(hl)
	set 2,(ix+dd)
	set 2,(iy+dd)
	set 2,a
	set 2,b
	set 2,c
	set 2,d
	set 2,e
	set 2,h
	set 2,l
	set 3,(hl)
	set 3,(ix+dd)
	set 3,(iy+dd)
	set 3,a
	set 3,b
	set 3,c
	set 3,d
	set 3,e
	set 3,h
	set 3,l
	set 4,(hl)
	set 4,(ix+dd)
	set 4,(iy+dd)
	set 4,a
	set 4,b
	set 4,c
	set 4,d
	set 4,e
	set 4,h
	set 4,l
	set 5,(hl)
	set 5,(ix+dd)
	set 5,(iy+dd)
	set 5,a
	set 5,b
	set 5,c
	set 5,d
	set 5,e
	set 5,h
	set 5,l
	set 6,(hl)
	set 6,(ix+dd)
	set 6,(iy+dd)
	set 6,a
	set 6,b
	set 6,c
	set 6,d
	set 6,e
	set 6,h
	set 6,l
	set 7,(hl)
	set 7,(ix+dd)
	set 7,(iy+dd)
	set 7,a
	set 7,b
	set 7,c
	set 7,d
	set 7,e
	set 7,h
	set 7,l

	sla (hl)
	sla (ix+dd)
	sla (iy+dd)
	sla a
	sla b
	sla c
	sla d
	sla e
	sla h
	sla l
	
	sra (hl)
	sra (ix+dd)
	sra (iy+dd)
	sra a
	sra b
	sra c
	sra d
	sra e
	sra h
	sra l
	
	srl (hl)
	srl (ix+dd)
	srl (iy+dd)
	srl a
	srl b
	srl c
	srl d
	srl e
	srl h
	srl l
	
	sub (hl)
	sub (ix+dd)
	sub (iy+dd)
	sub a
	sub b
	sub c
	sub d
	sub e
	sub h
	sub l
	sub n
	
	xor (hl)
	xor (ix+dd)
	xor (iy+dd)
	xor a
	xor b
	xor c
	xor d
	xor e
	xor h
	xor l
	xor n
