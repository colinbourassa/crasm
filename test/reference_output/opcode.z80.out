:104000008EDD8E23FD8E238F88898A8B8C8DCE2525
:10401000ED4AED5AED6AED7A86DD8623FD8623872B
:10402000808182838485C62509192939DD09DD1936
:10403000DD29DD39FD09FD19FD29FD39A6DDA623A0
:10404000FDA623A7A0A1A2A3A4A5E625CB46DDCB70
:104050002346FDCB2346CB47CB40CB41CB42CB4382
:10406000CB44CB45CB4EDDCB234EFDCB234ECB4FAC
:10407000CB48CB49CB4ACB4BCB4CCB4DCB56DDCBF6
:104080002356FDCB2356CB57CB50CB51CB52CB53E2
:10409000CB54CB55CB5EDDCB235EFDCB235ECB5F1C
:1040A000CB58CB59CB5ACB5BCB5CCB5DCB66DDCB56
:1040B0002366FDCB2366CB67CB60CB61CB62CB6342
:1040C000CB64CB65CB6EDDCB236EFDCB236ECB6F8C
:1040D000CB68CB69CB6ACB6BCB6CCB6DCB76DDCBB6
:1040E0002376FDCB2376CB77CB70CB71CB72CB73A2
:1040F000CB74CB75CB7EDDCB237EFDCB237ECB7FFC
:10410000CB78CB79CB7ACB7BCB7CCB7DCD3930DCFC
:104110003930FC3930D43930C43930F43930EC39E5
:1041200030E43930CC39303FBEDDBE23FDBE23BF85
:10413000B8B9BABBBCBDFE25EDA9ED89EDA1EDB1C5
:104140002F2735DD3523FD35233D050B0D151B1DB3
:10415000252BDD2BFD2B2D3BF310FEFBE3DDE3FDDB
:10416000E308EBD976ED46ED56ED5EED78DB25ED17
:1041700040ED48ED50ED58ED60ED6834DD3423FD41
:1041800034233C04030C14131C2423DD23FD232CB3
:1041900033EDAAEDBAEDA2EDB2E9DDE9FDE9C339EF
:1041A00030DA3930FA3930D23930C23930F2393078
:1041B000EA3930E23930CA3930380018FE30FC2094
:1041C000FA28F80212777071727374753625DD77EC
:1041D00023DD7023DD7123DD7223DD7323DD742382
:1041E000DD7523DD362325FD7723FD7023FD712347
:1041F000FD7223FD7323FD7423FD7523FD362325F6
:1042000032A05BED43A05BED53A05B22A05BDD22FF
:10421000A05BFD22A05BED73A05B0A1A7EDD7E230E
:10422000FD7E233AA05B7F78797A7B7CED577D3EDB
:1042300025ED5F46DD4623FD46234740414243448A
:10424000450625ED4BA05B01A05B4EDD4E23FD4EE8
:10425000234F48494A4B4C4D0E2556DD5623FD56FB
:1042600023575051525354551625ED5BA05B11A0B6
:104270005B5EDD5E23FD5E235F58595A5B5C5D1E0D
:104280002566DD6623FD66236760616263646526DB
:10429000252AA05B21A05BED47DD2AA05BDD21A0E4
:1042A0005BFD2AA05BFD21A05B6EDD6E23FD6E230E
:1042B0006F68696A6B6C6D2E25ED4FED7BA05BF925
:1042C000DDF9FDF931A05BEDA8EDB8EDA0EDB0EDA5
:1042D0004400B6DDB623FDB623B7B0B1B2B3B4B572
:1042E000F625EDBBEDB3ED79ED41ED49ED51ED591D
:1042F000ED61ED69D325EDABEDA3F1C1D1E1DDE1D8
:10430000FDE1F5C5D5E5DDE5FDE5CB86DDCB238615
:10431000FDCB2386CB87CB80CB81CB82CB83CB8459
:10432000CB85CB8EDDCB238EFDCB238ECB8FCB8865
:10433000CB89CB8ACB8BCB8CCB8DCB96DDCB23960D
:10434000FDCB2396CB97CB90CB91CB92CB93CB94B9
:10435000CB95CB9EDDCB239EFDCB239ECB9FCB98D5
:10436000CB99CB9ACB9BCB9CCB9DCBA6DDCB23A66D
:10437000FDCB23A6CBA7CBA0CBA1CBA2CBA3CBA419
:10438000CBA5CBAEDDCB23AEFDCB23AECBAFCBA845
:10439000CBA9CBAACBABCBACCBADCBB6DDCB23B6CD
:1043A000FDCB23B6CBB7CBB0CBB1CBB2CBB3CBB479
:1043B000CBB5CBBEDDCB23BEFDCB23BECBBFCBB8B5
:1043C000CBB9CBBACBBBCBBCCBBDC9D8F8D0C0F036
:1043D000E8E0C8ED4DED45CB16DDCB2316FDCB2334
:1043E00016CB17CB10CB11CB12CB13CB14CB15178D
:1043F000CB06DDCB2306FDCB2306CB07CB00CB01C1
:10440000CB02CB03CB04CB0507ED6FCB1EDDCB235B
:104410001EFDCB231ECB1FCB18CB19CB1ACB1BCB2E
:104420001CCB1D1FCB0EDDCB230EFDCB230ECB0FE4
:10443000CB08CB09CB0ACB0BCB0CCB0D0FED67C751
:10444000CFD7DFE7EFF7FF9EDD9E23FD9E239F98EA
:10445000999A9B9C9DDE25ED42ED52ED62ED7237FF
:10446000CBC6DDCB23C6FDCB23C6CBC7CBC0CBC1D0
:10447000CBC2CBC3CBC4CBC5CBCEDDCB23CEFDCB08
:1044800023CECBCFCBC8CBC9CBCACBCBCBCCCBCD20
:10449000CBD6DDCB23D6FDCB23D6CBD7CBD0CBD140
:1044A000CBD2CBD3CBD4CBD5CBDEDDCB23DEFDCB78
:1044B00023DECBDFCBD8CBD9CBDACBDBCBDCCBDD70
:1044C000CBE6DDCB23E6FDCB23E6CBE7CBE0CBE1B0
:1044D000CBE2CBE3CBE4CBE5CBEEDDCB23EEFDCBE8
:1044E00023EECBEFCBE8CBE9CBEACBEBCBECCBEDC0
:1044F000CBF6DDCB23F6FDCB23F6CBF7CBF0CBF120
:10450000CBF2CBF3CBF4CBF5CBFEDDCB23FEFDCB57
:1045100023FECBFFCBF8CBF9CBFACBFBCBFCCBFD0F
:10452000CB26DDCB2326FDCB2326CB27CB20CB21CF
:10453000CB22CB23CB24CB25CB2EDDCB232EFDCB07
:10454000232ECB2FCB28CB29CB2ACB2BCB2CCB2D5F
:10455000CB3EDDCB233EFDCB233ECB3FCB38CB390F
:10456000CB3ACB3BCB3CCB3D96DD9623FD962397B8
:10457000909192939495D625AEDDAE23FDAE23AFF8
:08458000A8A9AAABACADEE2521
:00000001FF
